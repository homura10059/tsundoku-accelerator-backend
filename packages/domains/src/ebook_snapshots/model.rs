use url::Url;

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}
