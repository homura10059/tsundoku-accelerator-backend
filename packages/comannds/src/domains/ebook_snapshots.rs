pub mod repositories;

use crate::domains::notifications::send_alert_message;
use anyhow::Result;
use db_client::prisma::PrismaClient;
use headless_chrome::Browser;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashSet;
use url::Url;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = match repositories::get(browser, id.as_str()) {
        Ok(s) => s,
        Err(e) => {
            let msg = format!("error snap_ebook id: {},  error:{}", id, e.to_string());
            send_alert_message(msg.clone()).await.unwrap();
            panic!("{}", msg);
        }
    };
    repositories::insert(client, &snapshot).await?;
    Ok(())
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

static PRICE_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"￥(?P<price>\d{1,3}(,\d{3})*)").unwrap());
static POINT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?P<points>\d{1,3}(,\d{3})*)pt").unwrap());

impl Payment {
    fn extract_price(text: &str) -> HashSet<&str> {
        let matched = PRICE_REGEX
            .captures_iter(text)
            .map(|cap| cap.name("price").unwrap().as_str())
            .collect::<Vec<_>>();
        matched.into_iter().collect()
    }

    fn extract_points(text: &str) -> HashSet<&str> {
        let matched = POINT_REGEX
            .captures_iter(text)
            .map(|cap| cap.name("points").unwrap().as_str())
            .collect::<Vec<_>>();
        matched.into_iter().collect()
    }

    pub fn new<T: AsRef<str>>(text: T) -> Option<Self> {
        let price = *Payment::extract_price(text.as_ref()).iter().next()?;
        let points = *Payment::extract_points(text.as_ref()).iter().next()?;
        Some(Payment {
            price: price.to_string().trim().replace(',', ""),
            points: points.to_string().trim().replace(',', ""),
        })
    }
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use db_client::prisma;
    use dotenv;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
    }

    #[test]
    fn test_payment_new() {
        let actual = Payment::new(
            r#"
            Kindle版 (電子書籍)
            ￥3,344
            獲得ポイント: 152pt
            "#,
        );
        let expected = Some(Payment {
            price: String::from("3344"),
            points: String::from("152"),
        });
        assert_eq!(actual, expected);
    }
}
