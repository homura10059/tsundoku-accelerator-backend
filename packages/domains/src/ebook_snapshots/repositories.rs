pub mod db;
pub mod scraper;
