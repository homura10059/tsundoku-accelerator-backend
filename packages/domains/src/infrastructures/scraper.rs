use crate::item_metadata::ItemMetaData;
use crate::models::WishListSnapshot;
use anyhow::Result;
use headless_chrome::{Browser, Element};
use pure_funcs::get_now_in_sec;
use url::Url;

pub fn search_from(attributes: &Vec<String>, key: &str) -> Option<String> {
    let target = attributes.iter().position(|attr| attr == key)? + 1;
    let (_pos, attr) = attributes
        .iter()
        .enumerate()
        .find(|(pos, _attr)| pos.eq(&target))?;
    let result = attr.clone();
    Some(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn search_from_returns_some() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "key_1");
        assert_eq!(actual, Some("val_1".to_string()))
    }

    #[test]
    fn search_from_returns_none() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "key_999");
        assert_eq!(actual, None)
    }

    #[test]
    fn search_from_returns_none2() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "val_1");
        assert_eq!(actual, None)
    }
}
