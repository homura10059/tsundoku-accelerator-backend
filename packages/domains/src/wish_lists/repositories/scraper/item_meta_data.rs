use crate::models::ItemMetaData;
use anyhow::Result;
use url::Url;

fn create_item_url(href: String) -> Result<Url> {
    let url = Url::parse("https://www.amazon.co.jp")?;
    let mut joined = url.join(href.as_str())?;
    joined.set_query(None);
    Ok(joined)
}

pub fn create(href: String, title: String, price: String) -> Result<ItemMetaData> {
    let url = create_item_url(href)?;
    let path = url.path().to_string();
    let a: Vec<_> = path.split("/").collect();
    let id = a.get(2).unwrap();
    let meta = ItemMetaData {
        id: id.to_string(),
        url,
        title,
        price,
    };
    Ok(meta)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_item_url() {
        assert_eq!(
            create_item_url("/dp/2BDAPI9RQ09E9/?coliid=IH".to_string()).unwrap(),
            Url::parse("https://www.amazon.co.jp/dp/2BDAPI9RQ09E9/").unwrap()
        );
    }

    #[test]
    fn test_create() {
        let expected = ItemMetaData {
            id: String::from("2BDAPI9RQ09E9"),
            url: Url::parse("https://www.amazon.co.jp/dp/2BDAPI9RQ09E9/").unwrap(),
            title: String::from("title"),
            price: String::from("100"),
        };
        assert_eq!(
            create(
                "/dp/2BDAPI9RQ09E9/?coliid=IH".to_string(),
                String::from("title"),
                String::from("100")
            )
            .unwrap(),
            expected
        );
    }
}
