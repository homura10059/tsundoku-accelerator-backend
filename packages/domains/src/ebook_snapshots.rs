pub mod repositories;

use anyhow::Result;
use headless_chrome::Browser;
use infrastructures::prisma::PrismaClient;
use url::Url;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = repositories::scraper::get(browser, id.as_str())?;
    repositories::db::insert(client, &snapshot).await?;
    Ok(())
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;
    use infrastructures::prisma;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        let actual = snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
        assert_eq!(actual, ());
    }
}
