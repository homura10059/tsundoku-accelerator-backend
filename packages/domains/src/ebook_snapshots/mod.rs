mod model;
pub mod repositories;

use anyhow::Result;
use headless_chrome::Browser;
use infrastructures::prisma::PrismaClient;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = repositories::scraper::get(browser, id.as_str())?;
    repositories::db::insert(client, &snapshot).await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;
    use infrastructures::prisma;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        let actual = snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
        assert_eq!(actual, ());
    }
}
