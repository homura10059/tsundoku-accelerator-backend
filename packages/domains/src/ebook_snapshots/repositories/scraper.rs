use crate::ebook_snapshots::model::{EbookSnapshot, Payment};
use anyhow::Result;
use chrono::Utc;
use headless_chrome::Browser;
use infrastructures::scraper;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashSet;
use url::Url;

static PRICE_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"￥(?P<price>\d{1,3}(,\d{3})*)").unwrap());
static POINT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?P<points>\d{1,3}(,\d{3})*)pt").unwrap());

pub fn create_url(id: &str) -> Result<Url> {
    let url = Url::parse("https://www.amazon.co.jp/dp/")?;
    let joined = url.join(id)?;
    Ok(joined)
}

fn extract_price(text: &str) -> HashSet<&str> {
    let matched = PRICE_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("price").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_points(text: &str) -> HashSet<&str> {
    let matched = POINT_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("points").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_payment(text: &str) -> Option<Payment> {
    let price = *extract_price(text).iter().next()?;
    let points = *extract_points(text).iter().next()?;
    Some(Payment {
        price: price.to_string().trim().replace(',', ""),
        points: points.to_string().trim().replace(',', ""),
    })
}

pub fn get(browser: &Browser, id: &str) -> Result<EbookSnapshot> {
    let url = create_url(id)?;

    let tab = browser.new_tab()?;
    tab.navigate_to(url.as_str())?;
    tab.wait_for_element("#navFooter")?;

    let image = tab.find_element("#ebooksImgBlkFront")?;
    let image_attribute = image.get_attributes()?.unwrap();
    let thumbnail_url_str = scraper::search_from(&image_attribute, "src").unwrap();
    let thumbnail_url = Url::parse(thumbnail_url_str.as_str())?;

    let price_list = tab.find_elements("#tmmSwatches .a-button-text")?;
    let payments = price_list
        .iter()
        .flat_map(|price| price.get_inner_text())
        .flat_map(|text| {
            let payment = extract_payment(text.as_str());
            payment
        })
        .collect::<Vec<_>>();

    let snapshot = EbookSnapshot {
        ebook_id: id.to_string(),
        scraped_at: Utc::now().timestamp(),
        thumbnail_url,
        payment_ebook: payments.get(0).cloned(),
        payment_real: payments.get(1).cloned(),
    };

    tab.close(true)?;

    Ok(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_get() {
        let browser = Browser::default().unwrap();
        let id = String::from("B09RQGMYKZ");
        let actual = get(&browser, id.as_str()).unwrap();
        assert_eq!(actual.ebook_id, id);
        assert_debug_snapshot!(actual.payment_ebook);
        assert_debug_snapshot!(actual.payment_real);
    }

    #[test]
    fn test_extract_payment() {
        let actual = extract_payment(
            r#"
            Kindle版 (電子書籍)
            ￥3,344
            獲得ポイント: 152pt
            "#,
        );
        let expected = Some(Payment {
            price: String::from("3344"),
            points: String::from("152"),
        });
        assert_eq!(actual, expected);
    }
}
