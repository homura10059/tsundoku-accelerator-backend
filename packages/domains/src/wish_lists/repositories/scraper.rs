use crate::infrastructures::scraper;
use crate::item_metadata::ItemMetaData;
use crate::models::WishListSnapshot;
use anyhow::Result;
use headless_chrome::{Browser, Element};
use pure_funcs::get_now_in_sec;
use url::Url;

fn create_url(id: &str) -> Result<Url> {
    let url = Url::parse("https://www.amazon.jp/hz/wishlist/ls/")?;
    let joined = url.join(id)?;
    Ok(joined)
}

fn get_item(elm: &Element) -> anyhow::Result<ItemMetaData> {
    let a_tag = elm.find_element(".a-link-normal")?;
    let a_tag_attributes = a_tag.get_attributes()?.unwrap();
    let href = scraper::search_from(&a_tag_attributes, "href").unwrap();
    let title = scraper::search_from(&a_tag_attributes, "title").unwrap();

    let attributes = elm.get_attributes()?.unwrap();
    let price = scraper::search_from(&attributes, "data-price").unwrap();

    let meta = ItemMetaData::new(href, title, price)?;
    Ok(meta)
}

pub fn get_wish_list_snapshot(id: &str) -> anyhow::Result<WishListSnapshot> {
    let url = create_url(id)?;
    let browser = Browser::default()?;

    let tab = browser.wait_for_initial_tab()?;
    tab.navigate_to(url.as_str())?;

    let nav_to_top = tab.wait_for_element("#navBackToTop")?;

    while tab.find_element("#endOfListMarker").is_err() {
        nav_to_top.scroll_into_view()?;
        break;
    }

    let selector_string = format!("[data-id=\"{}\"]", &id);
    let selector = selector_string.as_str();
    let item_list = tab.find_elements(selector)?;
    let mut items: Vec<_> = item_list
        .iter()
        .map(|item| get_item(item))
        .filter_map(|x| x.ok())
        .collect();

    items.sort();
    items.dedup();

    let title_element = tab.find_element("#profile-list-name")?;
    let title = title_element.get_inner_text()?;

    let snapshot = WishListSnapshot {
        id: id.to_string(),
        url,
        scraped_at: get_now_in_sec(),
        title,
        items,
    };

    Ok(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use url::Url;

    #[test]
    fn test_create_url() {
        assert_eq!(
            create_url("2BDAPI9RQ09E9").unwrap(),
            Url::parse("https://www.amazon.jp/hz/wishlist/ls/2BDAPI9RQ09E9").unwrap()
        );
    }

    #[test]
    fn test_get_item_urls() {
        let id = String::from("2BDAPI9RQ09E9");
        let url = Url::parse("https://www.amazon.jp/hz/wishlist/ls/2BDAPI9RQ09E9").unwrap();
        let actual = get_wish_list_snapshot(id.as_str()).unwrap();
        assert_eq!(actual.id, id);
        assert_eq!(actual.url, url);
        assert_eq!(actual.title, String::from("do_not_delete"));
        insta::assert_debug_snapshot!(actual.items);
    }
}
