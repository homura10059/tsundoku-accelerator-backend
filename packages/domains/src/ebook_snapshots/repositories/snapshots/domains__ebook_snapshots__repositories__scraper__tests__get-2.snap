---
source: domains/src/ebook_snapshots/repositories/scraper.rs
expression: actual.payment_real
---
Some(
    Payment {
        price: "3520",
        points: "106",
    },
)
